UnitystationLauncher/ContentScanning/AssemblyTypeChecker.cs

```diff
@@ -0,0 +1,1211 @@
+        ConcurrentBag<SandboxError> errors)
+    {
+        // This inheritance whitelisting primarily serves to avoid content doing funny stuff
+        // by e.g. inheriting Type.
+        foreach (var (_, baseType, interfaces) in inherited)
+        {
+            if (CanInherit(baseType) == false)
+            {
+                errors.Add(new SandboxError($"Inheriting of type not allowed: {baseType}"));
+            }
+
+            foreach (var @interface in interfaces)
+            {
+                if (CanInherit(@interface) == false)
+                {
+                    errors.Add(new SandboxError($"Implementing of interface not allowed: {@interface}"));
+                }
+            }
+
+            bool CanInherit(MType inheritType)
+            {
+                var realBaseType = inheritType switch
+                {
+                    ScanningTypes.MTypeGeneric generic => (ScanningTypes.MTypeReferenced)generic.GenericType,
+                    ScanningTypes.MTypeReferenced referenced => referenced,
+                    _ => throw new InvalidOperationException() // Can't happen.
+                };
+
+                if (IsTypeAccessAllowed(sandboxConfig, realBaseType, out var cfg) == false)
+                {
+                    return false;
+                }
+
+                return cfg.Inherit != InheritMode.Block && (cfg.Inherit == InheritMode.Allow || cfg.All);
+            }
+        }
+    }
+
+    private bool IsTypeAccessAllowed(SandboxConfig sandboxConfig, ScanningTypes.MTypeReferenced type,
+        [NotNullWhen(true)] out TypeConfig? cfg)
+    {
+        if (type.Namespace == null)
+        {
+            if (type.ResolutionScope is ScanningTypes.MResScopeType parentType)
+            {
+                if (IsTypeAccessAllowed(sandboxConfig, (ScanningTypes.MTypeReferenced)parentType.Type, out var parentCfg) == false)
+                {
+                    cfg = null;
+                    return false;
+                }
+
+                if (parentCfg.All)
+                {
+                    // Enclosing type is namespace-whitelisted so we don't have to check anything else.
+                    cfg = TypeConfig.DefaultAll;
+                    return true;
+                }
+
+                // Found enclosing type, checking if we are allowed to access this nested type.
+                // Also pass it up in case of multiple nested types.
+                if (parentCfg.NestedTypes != null && parentCfg.NestedTypes.TryGetValue(type.Name, out cfg))
+                {
+                    return true;
+                }
+
+                cfg = null;
+                return false;
+            }
+
+            if (type.ResolutionScope is ScanningTypes.MResScopeAssembly mResScopeAssembly &&
+                sandboxConfig.MultiAssemblyOtherReferences.Contains(mResScopeAssembly.Name))
+            {
+                cfg = TypeConfig.DefaultAll; //TODO DEBUG!!!
```

Comment:
Does this TODO still need done, or can it be removed?

Reply:
it should be okay, to check is above my skill level though a little bit